chrome/content/zotero/containers/collectionTree.jsx

```diff
@@ -0,0 +1,2179 @@
+					await collection.addItems(ids);
+				}.bind(this));
+			}
+			else if (targetTreeRow.isPublications()) {
+				await Zotero.Items.addToPublications(newItems, copyOptions);
+			}
+			
+			// If moving, remove items from source collection
+			if (dropEffect == 'move' && toMove.length) {
+				if (!sameLibrary) {
+					throw new Error("Cannot move items between libraries");
+				}
+				if (!sourceTreeRow || !sourceTreeRow.isCollection()) {
+					throw new Error("Drag source must be a collection for move action");
+				}
+				await Zotero.DB.executeTransaction(async function () {
+					await sourceTreeRow.ref.removeItems(toMove);
+				}.bind(this));
+			}
+		}
+		else if (dataType == 'text/x-moz-url' || dataType == 'application/x-moz-file') {
+			var targetLibraryID = targetTreeRow.ref.libraryID;
+			if (targetTreeRow.isCollection()) {
+				var parentCollectionID = targetTreeRow.ref.id;
+			}
+			else {
+				var parentCollectionID = false;
+			}
+			var addedItems = [];
+			
+			for (var i=0; i<data.length; i++) {
+				var file = data[i];
+
+				let item;
+				if (dataType == 'text/x-moz-url') {
+					var url = data[i];
+					
+					if (url.indexOf('file:///') == 0) {
+						let win = Services.wm.getMostRecentWindow("navigator:browser");
+						// If dragging currently loaded page, only convert to
+						// file if not an HTML document
+						if (win.content.location.href != url ||
+								win.content.document.contentType != 'text/html') {
+							var nsIFPH = Components.classes["@mozilla.org/network/protocol;1?name=file"]
+									.getService(Components.interfaces.nsIFileProtocolHandler);
+							try {
+								var file = nsIFPH.getFileFromURLSpec(url);
+							}
+							catch (e) {
+								Zotero.debug(e);
+							}
+						}
+					}
+					
+					// Still string, so remote URL
+					if (typeof file == 'string') {
+						let win = Services.wm.getMostRecentWindow("navigator:browser");
+						win.ZoteroPane.addItemFromURL(url, 'temporaryPDFHack', null, row); // TODO: don't do this
+						continue;
+					}
+					
+					// Otherwise file, so fall through
+				}
+				
+				if (dropEffect == 'link') {
+					item = await Zotero.Attachments.linkFromFile({
+						file: file,
+						collections: parentCollectionID ? [parentCollectionID] : undefined
+					});
+				}
+				else {
+					item = await Zotero.Attachments.importFromFile({
+						file: file,
+						libraryID: targetLibraryID,
+						collections: parentCollectionID ? [parentCollectionID] : undefined
+					});
+					// If moving, delete original file
+					if (dragData.dropEffect == 'move') {
+						try {
+							file.remove(false);
+						}
+						catch (e) {
+							Components.utils.reportError("Error deleting original file " + file.path + " after drag");
+						}
+					}
+				}
+				
+				addedItems.push(item);
+			}
+			
+			// Automatically retrieve metadata for PDFs
+			Zotero.RecognizePDF.autoRecognizeItems(addedItems);
+		}
+	}
+	
+	setDropEffect(event, effect) {
```

Comment:
That's exactly the goal, partially because it is unclear which functions will be viable for reuse and partially because it is easier to work this way when one of the components is still written in XUL and the other in React.